### PR TITLE
Feature/geotf

### PR DESCRIPTION
**Pull request by michaelpantic**
New library for geodetic data. I name it geotf.
For reference: From now on, geoframes are coordinate frames handled by geotf. TF frames are handled by regular tf.

Core features:
- Supports all frames supported by GDAL (https://gdal.org/) which is pretty much everything in existance.
- Can seamlessly and transparently convert between multiple geoframes 
- Supports mapping between geo frames and tf frames (through a common frame)
- Has methods to publish geodetic locations as tf frames
- Configuration of frames through rosparam (as they are static).

Example querys that are possible:
- GPS WGS 84 position in ENU frame
- UTM coordinates in ENU frame
- Current camera position in GPS coordinates (through camera_tf -> enu_tf = enu_geo -> gps_geo)
- Any UTM coordinate in current end-effector frame
- Any GPS coordinate in current camera frame.
- and so on.

Needed configurations:
- See demo.launch for an example
- Basically you configure geoframes with their names and their definition.
- Definition can be of type:
  - ENUOrigin (lon, lat, alt)
  - Well known geocode (GCS), such as "WGS84"
  - UTM zone (e.g. 32 north)
  - EPSG code. see https://epsg.io. Well known examples: New swiss coordinates (CH1903+, EPSG code 2056, see https://epsg.io/2056). Less known example: Maupiti 83 frame for some island in the pacific, EPSG code 4692, https://epsg.io/4692. 
- If TF adapter are used, a specification of which geo frame equals which tf frame. This is usually the ENU frame in some sort.

Especially through the EPSG facility, we probably never need to implement any other geodetic coordinate frame ever again, as basically every frame is defined in that database.

Publishing TF frames:
- Can publish any geolocation as TF frame based on the geo frame only. See demo_node.

### CODE REVIEW

**Comment by michaelpantic** [comment]
FYI: @rikba @marco-tranzatto  @clanegge @tstastny

File: geotf/src/geodetic_converter.cc [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+  }
+  XmlRpc::XmlRpcValue yaml_raw_data;
+  nh.getParam("/geotf", yaml_raw_data);
+
+  XmlRpc::XmlRpcValue& frame_definitions = yaml_raw_data["Frames"];
+  for (auto it = frame_definitions.begin();
+       it != frame_definitions.end(); ++it) {
+
+    const std::string frame_name = it->first;
+    XmlRpc::XmlRpcValue& xmlnode = it->second;
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
    auto& xmlnode = it->second;
```

File: geotf/src/geodetic_converter.cc [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+    std::string frame_type = xmlnode["Type"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:21]:
```suggestion
    auto frame_type = xmlnode["Type"];
```

File: geotf/src/geodetic_converter.cc [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+      int code = xmlnode["Code"];
+      addFrameByEPSG(frame_name, code);
+
+    } else if (frame_type == "GCSCode") {
+      if (!xmlnode.hasMember("Code")) {
+        ROS_WARN_STREAM("[GeoTF] Ignoring frame " << frame_type
+                                                  << ": GCSCode needs Code setting.");
+        continue;
+      }
+      std::string code = xmlnode["Code"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:37]:
```suggestion
      auto code = xmlnode["Code"];
```

File: geotf/src/geodetic_converter.cc [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+      addFrameByGCSCode(frame_name, code);
+
+    } else if (frame_type == "UTM") {
+      if (!xmlnode.hasMember("Zone") || !xmlnode.hasMember("Hemisphere")) {
+        ROS_WARN_STREAM("[GeoTF] Ignoring frame " << frame_type
+                                                  << ": UTM needs Zone and Hemisphere setting.");
+        continue;
+      }
+      int zone = xmlnode["Zone"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:46]:
```suggestion
      auto zone = xmlnode["Zone"];
```

File: geotf/src/geodetic_converter.cc [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+      bool hemisphere = xmlnode["Hemisphere"] == "N";
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:47]:
```suggestion
      auto hemisphere = xmlnode["Hemisphere"] == "N";
```

File: geotf/src/geodetic_converter.cc [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+      addFramebyUTM(frame_name, zone, hemisphere);
+
+    } else if (frame_type == "ENUOrigin") {
+      if (!xmlnode.hasMember("LatOrigin") || !xmlnode.hasMember("LonOrigin")
+          || !xmlnode.hasMember("AltOrigin")) {
+        ROS_WARN_STREAM("[GeoTF] Ignoring frame " << frame_type
+                                                  << ": ENUOrigin needs LatOrigin, LonOrigin and AltOrigin setting.");
+        continue;
+      }
+      double latOrigin = xmlnode["LatOrigin"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:57]:
```suggestion
      auto latOrigin = xmlnode["LatOrigin"];
```

File: geotf/src/geodetic_converter.cc [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+      double lonOrigin = xmlnode["LonOrigin"];
+      double altOrigin = xmlnode["AltOrigin"];
+
+      addFrameByENUOrigin(frame_name, lonOrigin, latOrigin, altOrigin);
+    }
+  }
+
+  // Get TF Mapping
+  if (yaml_raw_data.hasMember("TF_Mapping")) {
+    XmlRpc::XmlRpcValue& tf_mapping = yaml_raw_data["TF_Mapping"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
    auto& tf_mapping = yaml_raw_data["TF_Mapping"];
```

File: geotf/src/geodetic_converter.cc [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+    std::string geo_tf = tf_mapping["GEO_TF"];
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:68]:
```suggestion
    auto geo_tf = tf_mapping["GEO_TF"];
```

File: geotf/src/geodetic_converter.cc (hunk 7 above)
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
    auto& tf_mapping = yaml_raw_data["TF_Mapping"];
```

File: geotf/src/geodetic_converter.cc [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+// see https://spatialreference.org/ref/epsg/
+// Example: CH1903+ has epsg id 2056
+//          https://spatialreference.org/ref/epsg/2056/
+bool GeodeticConverter::addFrameByEPSG(const std::string& name, const int& id) {
+  if (mappings_.count(name)) {
+    return false;
+  }
+
+  // Create Spatial Reference from well known code
+  OGRSpatialReferencePtr
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  auto
```

File: geotf/src/geodetic_converter.cc [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+
+// Add a frame by UTM Zone
+// zone is the UTM zone (e.g. switzerland is in Zone 32)
+// north is true for northern hemisphere zones.
+bool GeodeticConverter::addFramebyUTM(const std::string& name,
+                                      const uint zone,
+                                      const bool north) {
+
+  // Create Spatial Reference from well known code
+  OGRSpatialReferencePtr
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  auto
```

File: geotf/src/geodetic_converter.cc [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+
+// Creates a new ENU Frame with its origin at the given
+// Location (lon,lat, alt)
+// Where (lon,lat,alt) are defined w.r.t. WGS84
+bool GeodeticConverter::addFrameByENUOrigin(const std::string& name,
+                                            const double lon,
+                                            const double lat,
+                                            const double alt) {
+  // Create Spatial Reference from well known code
+  OGRSpatialReferencePtr
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  auto
```

File: geotf/src/geodetic_converter.cc [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+      "[GeoTF] Added ENUOrigin " << lon << "/" << lat << "/" << alt <<
+                                 " as frame " << name);
+  mappings_.insert(std::make_pair(name, spatial_ref));
+  return true;
+}
+
+bool GeodeticConverter::addFrameByWKT(const std::string& name,
+                                      const std::string& wktformat) {
+  // Create Spatial Reference from well known code
+  OGRSpatialReferencePtr
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  auto
```

File: geotf/src/geodetic_converter.cc [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+                                    const Eigen::Affine3d& input,
+                                    const std::string& tf_output_frame,
+                                    Eigen::Affine3d* output,
+                                    const ros::Time& time) {
+  if (!tf_mapping_) {
+    ROS_WARN("[GeoTf] No TF mapping defined, canceling convertToTf");
+    return false;
+  }
+
+  std::string tf_connection_frame = tf_mapping_->second;
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:289]:
```suggestion
  auto tf_connection_frame = tf_mapping_->second;
```

File: geotf/src/geodetic_converter.cc [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+  std::string geotf_connection_frame = tf_mapping_->first;
+  Eigen::Affine3d tf_connection_value;
```
Comment (marco-tranzatto) [inline geotf/src/geodetic_converter.cc:291]:
```suggestion
  auto tf_connection_value;
```

**Review by marco-tranzatto — APPROVED** [review]
Very nice new library! :1st_place_medal: 

Since I recently read again Scott's Bible about C++14, I'm leaving suggestions for "auto" everywhere :rofl:

**Comment by michaelpantic** [comment]
@marco-tranzatto Thanks for the review!! 
Good point with the auto! I think it's a good idea to use it.
I'll take the following pragmatic approach:
- I'll use auto whenever it is immediately clear from the context what type is used (i.e.. when it is already specified in the same line) or when type reaallly doesn't matter (like xml parsing)
- I won't use auto for simple types (int, bool, strings etc)

File: geotf/src/geodetic_converter.cc (hunk 1 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: :+1:

File: geotf/src/geodetic_converter.cc (hunk 3 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:37]: I think here std::string could even be needed for template deduction stuff - so I leave it.

File: geotf/src/geodetic_converter.cc (hunk 2 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:21]: I think here std::string could even be needed for template deduction stuff - so I leave it.

File: geotf/src/geodetic_converter.cc (hunk 4 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:46]: simple type

File: geotf/src/geodetic_converter.cc (hunk 5 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:47]: simple type

File: geotf/src/geodetic_converter.cc (hunk 6 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:57]: simple type + template deduction

File: geotf/src/geodetic_converter.cc (hunk 7 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: changed

File: geotf/src/geodetic_converter.cc (hunk 8 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:68]: simple type + deduction

File: geotf/src/geodetic_converter.cc (hunk 7 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: changed

File: geotf/src/geodetic_converter.cc (hunk 9 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: :+1:

File: geotf/src/geodetic_converter.cc (hunk 10 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: :+1

File: geotf/src/geodetic_converter.cc (hunk 12 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc]: :+1:

File: geotf/src/geodetic_converter.cc (hunk 13 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:289]: simple-ish type

File: geotf/src/geodetic_converter.cc (hunk 14 above)
Comment (michaelpantic) [reply geotf/src/geodetic_converter.cc:291]: that needs to stay ;-)

File: geotf/include/geotf/geodetic_converter.h [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+// Prominent example: WGS84
+  bool addFrameByGCSCode(const std::string& name,
+                         const std::string& gcscode);
+
+  // Creates a new ENU Frame with its origin at the given
+  // Location (lon,lat, alt)
+  // Where (lon,lat,alt) are defined w.r.t. WGS84
+  bool addFrameByENUOrigin(const std::string& name,
+                           double lon,
+                           double lat,
```
Comment (rikba) [inline geotf/include/geotf/geodetic_converter.h:55]: Isn't the "standard" order Latitude, Longitude, Altitude?

File: geotf/src/demo_node.cc [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+  // Initialize ETH mainbuilding based on UTM coordiantes for example.
+  eth_mainbuilding_utm << 465882.064, 5247094.385, 498.217;
+
+  // Output ETH mainbuilding in GPS frame
+  if (converter.canConvert("UTM", "GPS")) {
+    converter.convert("UTM", eth_mainbuilding_utm,
+                      "GPS", &eth_mainbuilding_gps);
+
+    ROS_WARN_STREAM("ETH Mainbuilding WGS84 = " << std::setprecision(16)
```
Comment (rikba) [inline geotf/src/demo_node.cc]:
```suggestion
    ROS_INFO_STREAM("ETH Mainbuilding WGS84 = " << std::setprecision(16)
```

File: geotf/src/demo_node.cc [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+  } else {
+    ROS_WARN_STREAM("Frames not loaded.");
+  }
+
+  // Output ETH mainbuilding in Swissgrid frame
+  if (converter.canConvert("UTM", "CH1903+")) {
+    converter.convert("UTM", eth_mainbuilding_utm,
+                      "CH1903+", &eth_mainbuilding_ch);
+
+    ROS_WARN_STREAM("ETH Mainbuilding CH1903+/LV95 = " << std::setprecision(16)
```
Comment (rikba) [inline geotf/src/demo_node.cc]:
```suggestion
    ROS_INFO_STREAM("ETH Mainbuilding CH1903+/LV95 = " << std::setprecision(16)
```

File: geotf/src/demo_node.cc [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+  } else {
+    ROS_WARN_STREAM("Frames not loaded.");
+  }
+
+  // Output ETH mainbuilding in ENU frame based on LEE terasse
+  if (converter.canConvert("UTM", "ENU_LEE")) {
+    converter.convert("UTM", eth_mainbuilding_utm,
+                      "ENU_LEE", &eth_mainbuilding_enu);
+
+    ROS_WARN_STREAM("ETH Mainbuilding in ENU Frame based on LEE Terasse = "
```
Comment (rikba) [inline geotf/src/demo_node.cc]:
```suggestion
    ROS_INFO_STREAM("ETH Mainbuilding in ENU Frame based on LEE Terasse = "
```

File: geotf/src/geodetic_converter.cc [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+// Adds a coordinate frame by its EPSG identifier
+// see https://spatialreference.org/ref/epsg/
+// Example: CH1903+ has epsg id 2056
+//          https://spatialreference.org/ref/epsg/2056/
+bool GeodeticConverter::addFrameByEPSG(const std::string& name, const int& id) {
+  if (mappings_.count(name)) {
+    return false;
+  }
+
+  // Create Spatial Reference from well known code
```
Comment (rikba) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  // Create Spatial Reference from EPSG ID
```

File: geotf/src/geodetic_converter.cc [hunk 20]
```diff
@@ -0,0 +1,436 @@
...
+}
+
+// Add a frame by UTM Zone
+// zone is the UTM zone (e.g. switzerland is in Zone 32)
+// north is true for northern hemisphere zones.
+bool GeodeticConverter::addFramebyUTM(const std::string& name,
+                                      const uint zone,
+                                      const bool north) {
+
+  // Create Spatial Reference from well known code
```
Comment (rikba) [inline geotf/src/geodetic_converter.cc]:
```suggestion
  // Create Spatial Reference from UTM zone and North–South divide
```

File: geotf/src/geodetic_converter.cc [hunk 21]
```diff
@@ -0,0 +1,436 @@
...
+bool GeodeticConverter::addFrameByENUOrigin(const std::string& name,
+                                            const double lon,
+                                            const double lat,
+                                            const double alt) {
+  // Create Spatial Reference from well known code
+  auto spatial_ref = std::make_shared<OGRSpatialReference>();
+
+  // ENU Frame based on GPS coordinates
+  spatial_ref->SetWellKnownGeogCS("WGS84");
+  spatial_ref->SetOrthographic(lat, lon, 0.0, 0.0);
```
Comment (rikba) [inline geotf/src/geodetic_converter.cc:177]: This is also defined as `latitude, longitude, ...` thus I would change the argument order of `addFrameByENUOrigin`

File: geotf/src/geodetic_converter.cc [hunk 22]
```diff
@@ -0,0 +1,436 @@
...
+
+  altitude_offsets_.insert(std::make_pair(name, alt));
+
+  ROS_INFO_STREAM(
+      "[GeoTF] Added ENUOrigin " << lon << "/" << lat << "/" << alt <<
```
Comment (rikba) [inline geotf/src/geodetic_converter.cc]:
```suggestion
      "[GeoTF] Added ENUOrigin " << lat << "/" << lon << "/" << alt <<
```

File: geotf/src/demo_node.cc [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+  Eigen::Affine3d utm_building_point(Eigen::Affine3d::Identity());
+  utm_building_point.translation().x() = 465727;
+  utm_building_point.translation().y() = 5247291;
+  utm_building_point.translation().z() = 489.619;
+  std::cout << converter.publishAsTf("UTM", utm_building_point, "CornerUTM") << std::endl;
+
+  // Publish TF Frame CornerGPS based on UTM coordinates
+  Eigen::Affine3d gps_building_point(Eigen::Affine3d::Identity());
+  gps_building_point.translation().x() = 8.54616;
+  gps_building_point.translation().y() = 47.37823;
```
Comment (rikba) [inline geotf/src/demo_node.cc]: Change order latitude longitude.

**Review by rikba — APPROVED** [review]
Brilliant! :earth_africa: 

I'd just change the latitude->longitude->altitude order because I think it's standard to name latitude first. But you can correct me.

**Comment by michaelpantic** [comment]
@rikba  Thanks much for your review!

The confusion with lat/lon even confused me during development.
Issue is that GDAL uses lon/lat/alt ordering for spherical systems such as WGS - here we want to swap x and y. For ENU/Projected systems, it uses the same ordering as we expect (e/n/u), so here we don't want to swap x and y.

Fixed it in this way (I also think that lat/lon/alt is better for spherical systems):
https://github.com/ethz-asl/geodetic_utils/blob/bee3e7b293a69fab871a8e8c8eb56637cf26afd0/geotf/src/geodetic_converter.cc#L253
Basically I swap order for geographic/spherical systems but otherwise not. 
Not 100% sure if IsGeographic is always correct, but for the frames I tested it works ;-)

**Comment by michaelpantic** [comment]
Also we're not the only ones confused by this :-)
https://wiki.osgeo.org/wiki/Axis_Order_Confusion

**Comment by rikba** [comment]
I see I opened a bigger dilemma here... Given CDAL has its clear convention you may have also sticked to x=lon, y=lat.

**Comment by rikba** [comment]
> Also we're not the only ones confused by this :-) https://wiki.osgeo.org/wiki/Axis_Order_Confusion

Ironically axis order for WGS84/EPSG:4326 has been swapped in GDAL>3.0 as in Ubuntu >20.04. It is now [lat first, lon second](https://gdal.org/tutorials/osr_api_tut.html#crs-and-axis-order). I account for that on the branch [devel/noetic](https://github.com/ethz-asl/geodetic_utils/tree/devel/noetic) by removing the swapping. To be aware off ;) @michaelpantic
